### PR TITLE
Non-critical fixes related to M12 IRS demo work

### PR DESCRIPTION
**Pull request by andr3ej**
* Friendlier error message when a service can't be instantiated
* Fix shutdown deadlock after failure to register with network map

This PR only fixes one of the more obvious shutdown deadlocks, the real problem is our reliance on shutdown hooks for what a dependency injection container can do in a deterministic way.

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 1]
```diff
@@ -300,12 +303,16 @@ abstract class AbstractNode(open val configuration: NodeConfiguration,
...
-        val service = ctor.newInstance(services)
-        cordappServices.putInstance(clazz, service)
-        log.info("Installed ${clazz.name} Corda service")
+    fun <T : SerializeAsToken> installCordaService(serviceClass: Class<T>): T {
+        serviceClass.requireAnnotation<CordaService>()
+        val constructor = serviceClass.getDeclaredConstructor(PluginServiceHub::class.java).apply { isAccessible = true }
+        val service = try {
+            constructor.newInstance(services)
+        } catch (e: InvocationTargetException) {
+            throw ServiceInstantiationException(serviceClass.name, e.cause)
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Why is this needed when the exception is logged by the caller with the service details?

Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: It's really ugly in the console, you get InvocationTargetException: null

Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Why not just re-throw `e.cause`?

Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: The failure that motivated this change is a non-specific NoSuchElementException("List is empty."), so that would be worse than what we have now

Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Would it? If we just tweaked the error message that's logged in `installCordaServices` to say "Installing Corda service ${it.name} failed due to an error:" and we just re-throw e.cause then the logging would be sufficient.

Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: It looks like this:
```
E 12:50:55 [main] internal.Node.installCordaServices - Unable to install Corda service net.corda.irs.api.NodeInterestRates$Oracle
 java.util.NoSuchElementException: List is empty.
        at kotlin.collections.CollectionsKt___CollectionsKt.first(_Collections.kt:178) ~[kotlin-stdlib-1.1.2.jar:1.1.2]
        at net.corda.irs.api.NodeInterestRates$Oracle.<init>(NodeInterestRates.kt:91) ~[irs-demo-0.13-SNAPSHOT.jar:?]
        ...
```
You can tell by inspecting the stacktrace that the problem was in the service constructor, but on first glance it doesn't look related to the service implementation.

**Comment by shamsasari** [comment]
Why can't we simply change the log message to be more clearer and avoid the need for a separate exception:
```
E 12:50:55 [main] internal.Node.installCordaServices - service net.corda.irs.api.NodeInterestRates$Oracle threw the following exception when attempting to install it:
 java.util.NoSuchElementException: List is empty.
        at kotlin.collections.CollectionsKt___CollectionsKt.first(_Collections.kt:178) ~[kotlin-stdlib-1.1.2.jar:1.1.2]
        at net.corda.irs.api.NodeInterestRates$Oracle.<init>(NodeInterestRates.kt:91) ~[irs-demo-0.13-SNAPSHOT.jar:?]
        ...```

This would make things as clear you'd need to.

**Comment by andr3ej** [comment]
installCordaService is also used in Simulation.kt which propagates the exception. Also I don't think it's a good idea to make installCordaService return normally if it didn't install the service. And I don't want to change the generic "Unable to install Corda service" message as it catches other things that might go wrong, in which case we don't want to blame the service itself.

**Comment by shamsasari** [comment]
Sorry, I was referring to logging that's done in installCordaServices. Obviously installCordaService wouldn't return normally if the service didn't install

**Comment by andr3ej** [comment]
I don't see how I can make that work correctly without retaining the custom exception - or did you want me to catch ITE there? In which case the code is only good until installCordaService uses reflection more than once.

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 2]
```diff
@@ -234,27 +234,24 @@ fun <DI : DriverDSLExposedInterface, D : DriverDSLInternalInterface, A> genericD
         coerce: (D) -> DI,
         dsl: DI.() -> A
 ): A {
-    var shutdownHook: ShutdownHook? = null
+    val shutdownHook = addShutdownHook(driverDsl::shutdown)
```
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: This was deliberately here to avoid side effects in the constructor

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 3]
```diff
@@ -299,33 +296,25 @@ fun <A> poll(
...
-            } catch (t: Throwable) {
-                resultFuture.setException(t)
-                return@task
-            }
-            if (result == null) {
-                schedulePoll()
-            } else {
-                resultFuture.set(result)
+            ErrorOr.catch(check).match({
+                if (it != null) {
```
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Perhaps use named parameters here to indicate what these closures do

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 4]
```diff
@@ -299,33 +296,25 @@ fun <A> poll(
...
+                    executorService.schedule(this, pollInterval.toMillis(), MILLISECONDS)
+                }
+            }) {
+                resultFuture.setException(it)
             }
-        }, pollInterval.toMillis(), MILLISECONDS)
+        }
     }
-    schedulePoll()
+    executorService.submit(task)
```
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Perhaps run this explicitly on the current thread to avoid the scheduler roundtrip in case the poll condition is already satisfied

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 5]
```diff
@@ -459,24 +448,41 @@ class DriverDSL(
 
     override fun shutdown() {
         _shutdownManager?.shutdown()
-        _executorService?.shutdownNow()
+        _executorService?.let {
+            val n = it.shutdownNow().size
```
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Hmm why do we need this? It's fine if the backing threads don't terminate immediately (and actually it often takes some time for them to terminate), this will just cause a 1second hang in those cases

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 4 above)
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: The reason I made this change is that one of the usages is rpc.start, which can hang. The new behaviour is less astonishing IMO - there is no indication that poll(executor, block) will run the block on the foreground thread. I'm open to restoring the pre-check as an optional thing, but I don't think it's worth it is as this is only used in tests and polling is inefficient already. And you can run the pre-check manually if you really want to. I'll add a comment.

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 3 above)
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: The code looks very cluttered if I do that, and anyone who hasn't seen match before can ctrl-hover to see its params.

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 2 above)
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: I don't follow (which constructor?) - this code was 3 lines further down, and I've moved it here so the system.exit case is consistent with the finally case, which runs shutdown even if the first line of driverDsl.start fails.

File: core/src/main/kotlin/net/corda/core/Futures.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Each future is passed into the given handler as soon as it becomes done.
+ * By default invocations of the handler do not run concurrently, pass in null for [lock] to allow them to do so.
+ * The handler should throw [ThenContext.thenAgain] if it does not want to affect the returned future.
+ * Otherwise the result of the handler is copied into the returned future.
+ * If the handler errors and the returned future is already done, the error is logged (unless it's thenAgain).
+ * If the handler returns normally and the returned future is already done, the handler value is silently discarded.
+ */
+fun <T> Iterable<ListenableFuture<*>>.then(lock: Any? = Any(), handler: ThenContext.(ListenableFuture<*>) -> T): ListenableFuture<T> {
```
Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/Futures.kt]: I've simplified this for the current usages, and consequently thenAgain is not needed for this PR as both usages happen to short-circuit after the first result. I'm not going to replace with invokeAny because:
* invokeAny requires an executor, but this way we can join any old futures including those produced by poll, so this is much more flexible. This code is a logical multi-future extension of our existing single-future "then" function
* invokeAny cancels all but the first future, which sweeps exceptions under the carpet, which makes diagnosing weird bugs much more time-consuming

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 5 above)
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: I must have had a good reason for this. Currently the timeout is never achieved, so I'd rather keep it in so that we notice right away if anyone introduces another shutdown deadlock.

**Comment by andr3ej** [comment]
@exFalso back to you

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 3 above)
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Ah my bad thought this was in DriverDSL

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 2 above)
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Ah my bad thought this was in DriverDSL

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 3 above)
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Right, but it's better to be verbose than having to right click. Terse code is hard to read

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 5 above)
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Hmm I tried this route before and it didn't work, especially with artemis executors (i needed something like this for a thread leak test and had to compromise, otherwise the test would have taken way too long). If a thread is hanging then a hang is much more apparent (and easier to debug) than a log message somewhere, especially if it's part of a test where we don't actually read error log lines.

File: core/src/main/kotlin/net/corda/core/Futures.kt (hunk 6 above)
Comment (exFalso) [inline core/src/main/kotlin/net/corda/core/Futures.kt]: This is looking better, but the handler may be run several times, potentially completing the returned future several times.

> invokeAny requires an executor
That's why i suggested adding a primitive that works on futures instead.

Usually there are two basic ways to compose futures:

`fun both(a: Future<A>, b: Future<B>): Future<Pair<A, B>>`
in java this is Futures.allAsList

and

`fun first(a: Future<A>, b: Future<B>): Future<Either<A, B>>`
no idea why this is missing from java

The one we need is the latter. To be more javaesque and match your api better this could be
`fun <T> Iterable<ListenableFuture<T>>.first(): ListenableFuture<T>`.

For reference see
* Haskell: search for `waitAny`/`waitEither`/`race` https://hackage.haskell.org/package/async-2.1.1.1/docs/Control-Concurrent-Async.html
* Ocaml: search for `choose` https://ocaml.janestreet.com/ocaml-core/108.00.02/doc/async_core/Deferred.html
* Rust: https://docs.rs/futures/0.1.14/futures/future/fn.select_all.html

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 4 above)
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Fair enough

File: core/src/main/kotlin/net/corda/core/Futures.kt (hunk 6 above)
Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/Futures.kt]: I've fixed the re-entrant case now, so the handler only runs when the result future isn't done (with tests). The plan for this code (in another PR) is to implement a replacement for allAsList (that doesn't hide exceptions/hangs) on top of it, so the future that wins won't always be the first one, so for that reason I'm keeping then() as the name.

The most general case I'm thinking of is
```
val f1 = SettableFuture.create<Boolean>()
val f2 = SettableFuture.create<Boolean>()
val f3 = SettableFuture.create<Boolean>()
val g = listOf(f1, f2, f3).then {
    // Implement short-circuiting conjunction or disjunction here.
}
```
but not in this PR.

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 3 above)
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Done

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 5 above)
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Newest changes:
* Blow up after 5 seconds (assume hang)
* Log a warning if it succeeded but took more than 500 millis, to help diagnose slow builds
There is no such warning in the build log that includes these changes - I reckon at this point in the driver, the only tasks left are stray polls that cancel immediately.

**Comment by andr3ej** [comment]
@exFalso back to you

File: test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 5 above)
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Again, this shouldn't work like this, and it especially shouldn't throw an exception after timeout. If the test hangs, that's fine, we can inspect the threads, it's easy to debug. Please remove this, this change will only make our jobs harder when there is a problem during driver shutdown

File: core/src/main/kotlin/net/corda/core/Futures.kt [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+ * As soon as a given future becomes done, the handler is invoked with that future as its argument.
+ * Invocations of the handler will not run concurrently, so it's safe for it to maintain state.
+ * The result of the handler is copied into the result future, and the handler isn't invoked again.
+ * If a given future errors after the result future is done, the error is automatically logged.
+ */
+fun <T> Iterable<ListenableFuture<*>>.then(handler: (ListenableFuture<*>) -> T): ListenableFuture<T> {
+    return ThenContextImpl(this, handler = handler)
+}
+
+internal class ThenContextImpl<T>(futures: Iterable<ListenableFuture<*>>, private val log: Logger = defaultLog, handler: (ListenableFuture<*>) -> T) : AbstractFuture<T>() {
```
Comment (exFalso) [inline core/src/main/kotlin/net/corda/core/Futures.kt]: Why is this file so complicated?? All we need is a construct that races futures and returns the first one! The signature should literally be `fun <T> Iterable<ListenableFuture<T>>.first(): ListenableFuture<T>`
